explainer.md

```diff
@@ -345,6 +341,114 @@ And here's the result:
+```js
+document.startViewTransition(() => updateDOMToNewPage());
+```
+
+This results in a cross-fade between the 2 Documents from the default CSS set up by the browser. The transition executes only if this tag is present on both the old and new Documents. The tag must also be added before the body element is parsed.
+
+The motivation for a declarative opt-in, instead of a script event, is:
+
+* Enabling authors to define transitions with no script. If the transition doesn't need to be customized based on the old/new URL, it can be defined completely in CSS.
+
+* Avoiding undue latency in the critical path for browser initiated navigations like back/forward. We want to avoid dispatch of a script event for each of these navigations.
+
+Issue: The meta tag can be used to opt-in to other navigation types going forward: same-document, same-site, etc.
+
+### Script Events
+
+Script can be used to customize a transition based on the URL of the old/new Document; or the current state of the Document when the transition is initiated. The Document could've been updated since first old from user interaction.
+
+#### Script on Old Document
+```js
+document.addEventListener("viewTransitionOnOldDocument", (event) => {
+  // Cancel the transition (based on new URL) if needed.
+  if (shouldNotTransition(event.url)) {
+    event.preventDefault();
+    return;
+  }
+    
+  // Set up names on elements based on the new URL.
+  if (shouldTagThumbnail(event.url))
+    thumbnail.style.viewTransitionName = "full-embed";
+    
+  // Add opaque contextual information to share with the new Document.
+  // This must be [serializable object](https://developer.mozilla.org/en-US/docs/Glossary/Serializable_object).
+  event.context = createTransitionContext(event.url);
+});
+```
+
+#### Script on New Document
+```js
+// This event must be registered before the `body` element is parsed.
+document.addEventListener("viewTransitionOnNewDocument", (event) => {
+  // Cancel the transition (based on old URL) if needed.
+  if (shouldNotTransition(event.url)) {
+    event.preventDefault();
+    return;
+  }
+  
+  // The `ViewTransition` object associated with this transition.
+  const transition = event.transition;
+  
+  // Retrieve the list of names captured on the old Document.
+  const oldNames = event.oldNames;
+  
+  // Retrieve the context generated by the old Document.
+  const context = event.context;
+  
+  // Add render-blocking resources to delay the first paint and transition
+  // start. This can be customized based on the old Document state when the
+  // transition was initiated.
+  markRenderBlockingResources(oldNames, context);
+  
+  // The browser defers painting the new Document until all render-blocked
+  // resources have been fetched or timed out.
+  // The `domUpdated` promise can be used to inspect when the above is
+  // done and add viewTransitionNames based on what the new Document's
+  // first paint would look like.
+  transition.domUpdated.then(() => {
```

Comment:
I don't think we should reuse `domUpdated` for this, as it doesn't really mean the same thing. We could add a new promise for it if needed. `transition` doesn't need to be the exact same type.

Reply:
I can see the benefit of pulling domUpdated into an independent API for render-blocking in general. But other than that, do the ready/finished promise make sense? They are conceptually the same and needed for the same use-case as the SPA API. I can move them into a different ViewTransition type.

Reply:
Removed domUpdated promise with an open issue for providing developers with a signal to detect when DOM is ready for first paint. So they could conditionally decide what is tagged.